### PR TITLE
Add a crypto handler

### PR DESCRIPTION
**Pull request by babolivier**
To handle requests with an Olm-encrypted body, see https://github.com/matrix-org/matrix-content-scanner-python/blob/main/docs/api.md#encrypted-post-body

~Mypy checks are currently failing, I'm trying to work that out at https://gitlab.matrix.org/matrix-org/olm/-/merge_requests/62~ fixed now

The PyPI version of `python-olm` is too out of date to have type annotations (and the type marker), and updating it is currently blocked behind https://github.com/vector-im/legal-compliance/issues/223 and friends (which is at least a month away from being resolved).

### CODE REVIEW

File: tox.ini
Comment (richvdh) [inline tox.ini:43]: it would be good to switch to olm's main repo, once https://gitlab.matrix.org/matrix-org/olm/-/merge_requests/62 is merged

File: matrix_content_scanner/crypto.py
Comment (richvdh) [inline matrix_content_scanner/crypto.py:56]: shouldn't this be logged *before* generating the new pickle?

File: matrix_content_scanner/crypto.py
Comment (richvdh) [inline matrix_content_scanner/crypto.py:59]: if `pickle_path` is invalid (eg it points to a non-existent dir, or an unwritable location)... will this write something sensible to stderr on startup?

File: matrix_content_scanner/crypto.py
Comment (richvdh) [inline matrix_content_scanner/crypto.py:81]: aren't you supposed to update the pickle after a decryption? (else what's the point of the pickle?)

**Review by richvdh — COMMENTED** [review]
seems plausible, I guess.

I can't help but feel that using libolm is a bit of a retrograde step nowadays: shouldn't we be using the rust `matrix-sdk-crypto` these days? Not that I have any real idea how much work it would be to create python bindings for that, but it might be worth running past the crypto folks for an opinion.

File: tox.ini
Comment (babolivier) [reply tox.ini:43]: The plan once that MR is merged is to replace this block with:

```ini
install_command = python -m pip install --index-url=https://gitlab.matrix.org/api/v4/projects/27/packages/pypi/simple {opts} {packages}
```

This will cause pip to fetch all packages from the gitlab-hosted pypi repo for olm. If a package cannot be found there (which will be the case for any package that isn't `python-olm`) GitLab will automatically forward the request to pypi.org according to https://gitlab.matrix.org/help/user/packages/pypi_repository/index#install-a-pypi-package (as of GitLab 14.2, and I've confirmed with ops that gitlab.matrix.org runs a more recent version of GitLab).

Longer term, the hope is to have a package on pypi.org (either for just the bindings, or the bindings + a build version of olm), but that's currently blocked on legal concerns which I'm told will take at least a month to resolve.

**Comment by babolivier** [comment]
> I can't help but feel that using libolm is a bit of a retrograde step nowadays: shouldn't we be using the rust `matrix-sdk-crypto` these days? Not that I have any real idea how much work it would be to create python bindings for that, but it might be worth running past the crypto folks for an opinion.

The long-term solution will be to use vodozemac, which will have first-party Python bindings: https://github.com/matrix-org/vodozemac-bindings/tree/main/python. However, currently these are very beta, undocumented, and don't exist on a pypi-style repo.

Vodozemac also currently does not provide (as far as I can tell) a replacement for the `olm.pk` module (which refers to [this part](https://gitlab.matrix.org/matrix-org/olm/-/blob/master/src/pk.cpp) of the libolm code), so it can't easily (if at all) be used as a drop-in replacement for libolm in the content scanner currently.

Another thing that is nice with using olm here is that it can directly use pickle files from https://github.com/matrix-org/matrix-content-scanner, whereas using vodozemac would require a migration step ([ref](https://docs.rs/vodozemac/latest/vodozemac/#pickling)). I think it would be best to implement this migration as a second step after the rewrite is done.

File: matrix_content_scanner/crypto.py
Comment (babolivier) [reply matrix_content_scanner/crypto.py:81]: Are you? My understanding of the `PkDecryption` class is that it's classic public-key encryption, thus the keys stay the same after decryption. This seems to correlate with the docstring of [`pk.py`](https://gitlab.matrix.org/matrix-org/olm/-/blob/master/python/olm/pk.py). So at this point the pickle is just a fancy way to store the private key, but also apparently the only way the bindings allow us to build the `PkDecryption` object.

File: matrix_content_scanner/config.py
Comment (richvdh) [inline matrix_content_scanner/config.py:211]: `_check_file_path` doesn't appear to create the file, in contrast to what the comment says.

In general: I don't think that separating the checks from the actual use of the file is good, for a number of reasons:
 * it tightly couples this config code to the way that the file is used. (For example: we have embedded the knowledge here that we only need read access to the file if it already exists. That's fine as far as it goes, but is ripe for getting out of sync if we change the way `CryptoHandler` works.)
 * You're kinda relying on `os.access` and `open` to interact in the way you expect. One example is that `os.access` checks with the *real* userid, whereas `open` uses the *effective* userid. Admittedly they are unlikely to be different, but it feels like an unnecessary assumption, and ties us quite hard to unix-like filesystems. Another example: the disk might be full (so `os.access` is happy, but you still can't write the file).
 * Just because the file exists/is writable here, doesn't necessarily mean that it will still be so when we come to use it later. For example, the file might get deleted in the meantime, or created with different permissions.

All in all: I think it would be better to make sure we construct `CryptoHandler` at startup, and handle the exceptions from `open` in a way that gives us nicer output.

File: matrix_content_scanner/crypto.py
Comment (richvdh) [inline matrix_content_scanner/crypto.py:61]: ok, so something that confused me on my first review: the logs and comments here talk about loading the pickle, but don't explain what that pickle is actually doing.

So, rather than taking about pickles, can we instead talk about private keys. "First try to load our keypair from the pickle file" ... "If the pickle file doesn't exist, create a new keypair and save it in a new pickle". (and update the log statement accordingly)

Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

(TBH I slightly wonder if silently creating a new key is a footgun. I might prefer to *know* that my keyfile has gone missing, rather than suddenly getting new keys?)

**Review by richvdh — CHANGES_REQUESTED** [review]

File: matrix_content_scanner/crypto.py
Comment (babolivier) [reply matrix_content_scanner/crypto.py:61]: > Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

Hah, I forgot to port that bit :facepalm: thanks for noticing!

> (TBH I slightly wonder if silently creating a new key is a footgun. I might prefer to know that my keyfile has gone missing, rather than suddenly getting new keys?)

fwiw the point of this is to mirror what the current content scanner does, but I guess the creation of the file can be moved to a script.

File: matrix_content_scanner/config.py
Comment (babolivier) [reply matrix_content_scanner/config.py:211]: > `_check_file_path` doesn't appear to create the file, in contrast to what the comment says.

Ah I phrased this poorly, it meant to read as "that we can [...] create it if it doesn't exist".

> I don't think that separating the checks from the actual use of the file is good, for a number of reasons:

Yeah good points. I'll try to rework that differently.

File: matrix_content_scanner/crypto.py
Comment (babolivier) [reply matrix_content_scanner/crypto.py:61]: > Also: where is the config file documented? It should say something about the fact a new key is magically created if the pickle doesn't exist.

Oh wait no, that bit does exist: https://github.com/matrix-org/matrix-content-scanner-python/blob/b7563c3e6e5a6fe98ea9578623b51e3a7a77f91f/config.sample.yaml#L101-L109

(but yeah it should mention if it's creating the file)

File: matrix_content_scanner/config.py
Comment (babolivier) [reply matrix_content_scanner/config.py:211]: > I think it would be better to make sure we construct CryptoHandler at startup, and handle the exceptions from open in a way that gives us nicer output.

It's worth noting that `CryptoHandler` _will_ be constructed at startup: it will be constructed by one of the `/..._encrypted` http resources (such as [this one](https://github.com/matrix-org/matrix-content-scanner-python-wip/blob/88ab968106537a1451e9b11771a5cd24df0b2376/matrix_content_scanner/servlets/download.py#L54)), which are constructed by the [`HTTPServer` constructor](https://github.com/matrix-org/matrix-content-scanner-python-wip/blob/main/matrix_content_scanner/httpserver.py#L34-L63). An `HTTPServer` is constructed on `MatrixContentScanner.start`.

File: tox.ini
Comment (babolivier) [reply tox.ini]: Olm 3.2.13 has been released with the change, so I've now put that plan in motion.

File: README.md
Comment (richvdh) [inline README.md:33]: mhmm, I'm not entirely sure about this approach. My experience with synapse (which does similar things with key files) is that requirements like this make it a bit of a pain to run in kubernetes-like environments, and led to https://github.com/matrix-org/synapse/pull/13615.

I realise this comment is in conflict with my earlier comment. I'm just not sure what is the best approach.

I suggest you stick with what you have, to avoid further vacillation. Just warning you that you might need to revisit this.

File: config.sample.yaml
Comment (richvdh) [inline config.sample.yaml:106]: comma splice!
```suggestion
    # for the Matrix Content Scanner to run; see README.md for more information.
```

File: matrix_content_scanner/mcs.py
Comment (richvdh) [inline matrix_content_scanner/mcs.py:142]: for the record, I'd generally recommend writing an error message directly to stderr (`print(..., file=sys.stderr)`) rather than indirecting via the logging system for problems encountered during startup, but up to you.

File: tox.ini
Comment (richvdh) [inline tox.ini:37]: I'm not sure there's any particular intention to maintain the packages on pypi?

**Review by richvdh — APPROVED** [review]
some random nits and comments but nothing significant

lgtm!

File: tox.ini
Comment (richvdh) [reply tox.ini:37]: well, apparently, It's Complicated (TM)

File: matrix_content_scanner/mcs.py
Comment (babolivier) [reply matrix_content_scanner/mcs.py:142]: Oh yes good point

File: README.md
Comment (babolivier) [reply README.md:33]: Yeah I'm also on the fence. I initially had it automatically generate secrets because that's what the javascript content scanner does, but there are good arguments in both directions tbh.

> I suggest you stick with what you have, to avoid further vacillation. Just warning you that you might need to revisit this.

Actually, the issue you linked above seems to indicate this approach would cause issues with k8s-style deployments. I know there's an intent to deploy the content scanner on EMS (and actually most of the infrastructure is already in place for this afaik) so I'd rather revert to generating it automatically.

**Comment by babolivier** [comment]
@richvdh Asking for another review despite the :heavy_check_mark: since I've made some changes that seem significant enough to warrant another review (with more context in https://github.com/matrix-org/matrix-content-scanner-python/pull/26#discussion_r992393882).

File: config.sample.yaml
Comment (richvdh) [inline config.sample.yaml]: we should say that it will be created if it doesn't exist.

File: matrix_content_scanner/crypto.py
Comment (richvdh) [inline matrix_content_scanner/crypto.py:65]: shouldn't this be:

```
except FileNotFoundError:
 ...
except OSError:
    raise ConfigError(...)
```

?

File: matrix_content_scanner/crypto.py
Comment (babolivier) [reply matrix_content_scanner/crypto.py:84]: Gah, indeed, for some reason I keep forgetting it's a thing.

File: matrix_content_scanner/crypto.py
Comment (babolivier) [reply matrix_content_scanner/crypto.py:65]: Oh yes totally

File: config.sample.yaml
Comment (richvdh) [inline config.sample.yaml:105]: I feel like this misses the important point somewhat.

```suggestion
    # A new keypair will be created at startup if the pickle file doesn't already exist.
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

**Pull request merged**
